server/hotrod/src/main/scala/org/infinispan/server/hotrod/ClientListenerRegistry.scala

```diff
@@ -99,7 +105,28 @@ class ClientListenerRegistry(configuration: HotRodServerConfiguration) extends L
       }
 
       eventSenders.put(listenerId, clientEventSender)
-      cache.addListener(clientEventSender, filter.orNull, converter.orNull)
+
+      if (includeState) {
+         // If state included, do it async
+         val cf = CompletableFuture.runAsync(new Runnable {
+            override def run(): Unit = {
+               cache.addListener(clientEventSender, filter.orNull, converter.orNull)
+            }
+         }, addListenerExecutor)
+
+         cf.whenComplete(new BiConsumer[Void, Throwable] {
+            override def accept(t: Void, cause: Throwable): Unit = {
+               val resp =
+                  if (cause == null) decoder.createSuccessResponse(h, null)
+                  else decoder.createErrorResponse(h, cause)
+
+               ch.writeAndFlush(resp)
+         }
+         })
+      } else {
+         cache.addListener(clientEventSender, filter.orNull, converter.orNull)
+         ch.writeAndFlush(decoder.createSuccessResponse(h, null))
+      }
```

Comment:
Cool, I'll sort those out. 

As a FYI: using Scala Futures required more tinkering when those were used inside WildFly, so given the short amount of time to get this fixed, decided to fallback on Java.